src/ui/shared/UrlContext.ts

```diff
@@ -0,0 +1,97 @@
+export type AppMode = 'onboarding' | 'wallet';
+export type WindowType = 'popup' | 'tab' | 'dialog';
+export type WindowLayout = 'column' | 'page';
+
+export enum UrlContextParam {
+  windowType = 'windowType',
+  windowLayout = 'windowLayout',
+  appMode = 'appMode',
+}
+
+function getUrlContextParam<T>(param: UrlContextParam, defaultValue?: T) {
+  const url = new URL(window.location.href);
+  return (url.searchParams.get(param) as T) || defaultValue;
+}
+
+export interface UrlContextParams {
+  appMode?: AppMode;
+  windowType?: WindowType;
+  windowLayout?: WindowLayout;
+}
+
+class UrlContext {
+  protected params: UrlContextParams;
+
+  constructor(params: UrlContextParams) {
+    this.params = params;
+  }
+
+  set(searchParams: URLSearchParams, params: UrlContextParams) {
+    if (params.windowLayout) {
+      searchParams.set(UrlContextParam.windowLayout, params.windowLayout);
+      this.params.windowLayout = params.windowLayout;
+    }
+    if (params.windowType) {
+      searchParams.set(UrlContextParam.windowType, params.windowType);
+      this.params.windowType = params.windowType;
+    }
+    if (params.appMode) {
+      searchParams.set(UrlContextParam.appMode, params.appMode);
+      this.params.appMode = params.appMode;
+    }
```

Comment:
I avoided using a for loop here, so there is some duplication instead of
```
for (const [key, value] of Object.entries(params)) {
  if (value) {
    ...
  }
```